### PR TITLE
ksm_ksmtuned: gives more time for ksm to start

### PR DESCRIPTION
**Pull request by mcasquer**
ksm_ksmtuned: gives more time for ksm to start

After checking the timings, the case failed because
ksmtuned needs some more time in order to start ksm
after reaching the threshold.

Signed-off-by: mcasquer <mcasquer@redhat.com>
ID: 1932

### CODE REVIEW

**Comment by mcasquer** [comment]
Checked 5 times, the current increase should be enough
```
 (1/5) repeat1.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (1/5) repeat1.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (344.12 s)
 (2/5) repeat2.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (2/5) repeat2.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (297.25 s)
 (3/5) repeat3.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (3/5) repeat3.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (340.82 s)
 (4/5) repeat4.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (4/5) repeat4.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (342.92 s)
 (5/5) repeat5.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (5/5) repeat5.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (293.97 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

**Comment by mcasquer** [comment]
@yanan-fu could you review this PR? Thanks !

**Comment by yanan-fu** [comment]
would we use the `wait_for` method to check the status before timeout ?

**Comment by mcasquer** [comment]
> would we use the `wait_for` method to check the status before timeout ?

@yanan-fu IINM, the `wait_for` function waits for the process to finish, but the idea is to wait **before** starting the process (check KSM status). Is there a different way to do it?

**Comment by yanan-fu** [comment]
> > would we use the `wait_for` method to check the status before timeout ?
> 
> @yanan-fu IINM, the `wait_for` function waits for the process to finish, but the idea is to wait **before** starting the process (check KSM status). Is there a different way to do it?

Not exactly,  it is wait for get the expected result instead of `process to finish`.    ksm status is running is the check point.

**Comment by mcasquer** [comment]
> > > would we use the `wait_for` method to check the status before timeout ?
> > 
> > 
> > @yanan-fu IINM, the `wait_for` function waits for the process to finish, but the idea is to wait **before** starting the process (check KSM status). Is there a different way to do it?
> 
> Not exactly, it is wait for get the expected result instead of `process to finish`. ksm status is running is the check point.

Sorry, it seems there are two wait_for functions :grinning:

**Comment by mcasquer** [comment]
Test passed with the new changes
```
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (308.47 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

**Comment by mcasquer** [comment]
@yanan-fu could you review again this PR? Thanks !

**Comment by mcasquer** [comment]
@yanan-fu this is a kindly reminder, could you review again this PR? Thanks !

File: qemu/tests/ksm_ksmtuned.py [hunk 1]
```diff
@@ -83,23 +82,24 @@ def check_qemu_used_mem(qemu_pid, mem):
                                    stress_timeout, 10, 10):
             test.error("QEMU used memory doesn't reach %s of guest mem %sM in "
                        "%ss" % (mem_thres, mem // 1024, stress_timeout))
-        time.sleep(30)
+        cmd = params['cmd_check_ksm_status']
         free_mem_host = utils_memory.freememtotal()
-        ksm_status = process.getoutput(params['cmd_check_ksm_status'])
+        ksm_status = utils_misc.wait_for(lambda: '1' == process.getoutput(cmd),
+                                         60)
```
Comment (yanan-fu) [inline qemu/tests/ksm_ksmtuned.py:88]: Base on your test result,  30s is not enough sometimes but can work most of the time, suggest add `first` param which means `Time to sleep before first attempt`, for example `first = 20`

Comment (mcasquer) [reply qemu/tests/ksm_ksmtuned.py:88]: @yanan-fu IIUC I need to add a sleep with the first value and reduce the wait_for till 30s again?

**Comment by mcasquer** [comment]
```
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (279.51 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

File: qemu/tests/ksm_ksmtuned.py (hunk 1 above)
Comment (yanan-fu) [reply qemu/tests/ksm_ksmtuned.py]: @mcasquer  I mean the `first` param for `wait_for` function

Comment (yanan-fu) [reply qemu/tests/ksm_ksmtuned.py]: refer:https://github.com/avocado-framework/avocado-vt/blob/2ddf56c099e314351d624bc33d6a7289f33a9804/virttest/utils_misc.py#L543

Comment (mcasquer) [reply qemu/tests/ksm_ksmtuned.py]: @yanan-fu when replacing the function call to:
`utils_misc.wait_for(lambda: '1' == process.getoutput(cmd), 40, first=20.0)`
The test case fails complaining about KSM should be running
```
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: FAIL: KSM should be running (258.96 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```
Any idea what's the difference that make the same test pass with:
`utils_misc.wait_for(lambda: '1' == process.getoutput(cmd), 60)`

Comment (mcasquer) [reply qemu/tests/ksm_ksmtuned.py:88]: After trying on a different host, timeout=40s and first=20.0 works fine

Comment (mcasquer) [reply qemu/tests/ksm_ksmtuned.py]:
```
 (1/5) repeat1.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (1/5) repeat1.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (290.78 s)
 (2/5) repeat2.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (2/5) repeat2.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (286.98 s)
 (3/5) repeat3.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (3/5) repeat3.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (298.66 s)
 (4/5) repeat4.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (4/5) repeat4.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (296.60 s)
 (5/5) repeat5.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: STARTED
 (5/5) repeat5.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.ksm_ksmtuned.q35: PASS (296.76 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

Comment (mcasquer) [reply qemu/tests/ksm_ksmtuned.py]: @yanan-fu do you agree with these values?

Comment (yanan-fu) [reply qemu/tests/ksm_ksmtuned.py]: Ok for me, thanks!

**Comment by yanan-fu** [comment]
Call for multi-arch review, @PaulYuuu @fbq815 @MiriamDeng , thanks ~

**Comment by MiriamDeng** [comment]
It seems there's error as below on ppc64le
 (2/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: STARTED
 (2/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: ERROR: QEMU used memory doesn't reach 0.95 of guest mem 37888.0M in 1800s (2242.90 s)

**Comment by yanan-fu** [comment]
> It seems there's error as below on ppc64le (2/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: STARTED (2/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: ERROR: QEMU used memory doesn't reach 0.95 of guest mem 37888.0M in 1800s (2242.90 s) 

Logically, it is not related with this patch, it is the `mem_thres` defined in the script and qemu does not reach it after stress i think.

**Comment by mcasquer** [comment]
> Logically, it is not related with this patch, it is the `mem_thres` defined in the script and qemu does not reach it after stress i think.

That's it, I am debugging the case on @MiriamDeng environment but for sure it is not related with the current patch. So that makes me think about how long is being ksm_ksmtuned failing on ppc... :thinking: :sweat_smile:

**Comment by mcasquer** [comment]
> > Logically, it is not related with this patch, it is the `mem_thres` defined in the script and qemu does not reach it after stress i think.
> 
> That's it, I am debugging the case on @MiriamDeng environment but for sure it is not related with the current patch. So that makes me think about how long is being ksm_ksmtuned failing on ppc... 🤔 😅

Furthermore, the test case takes ~5 minutes on x86_64 but more than 30 minutes in ppc

**Review by PaulYuuu — APPROVED** [review]

**Comment by mcasquer** [comment]
@fbq815 @MiriamDeng this is a kindly reminder, could you review this PR? Thanks !

**Comment by fbq815** [comment]
Test result on s390x:
RHEL9:
ksm_ksmtuned.s390-virtio: PASS (158.85 s)
RHEL8:
ksm_ksmtuned.s390-virtio: PASS (177.00 s)
LGTM

**Review by fbq815 — APPROVED** [review]

**Comment by MiriamDeng** [comment]
ACK
 (1/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: STARTED
 (1/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (2061.82 s)
 (2/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: STARTED
 (2/2) Host_RHEL.m8.u10.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: PASS (385.82 s)

**Pull request merged**